### PR TITLE
fix(compiler): don't throw when Symbol is used as DI token

### LINKED ISSUE #13314: Cannot create property '__anonymousType' on symbol
**I'm submitting a ...**  (check one with "x")
```
[x] bug report
[ ] feature request
[ ] support request
```

**Current behavior**
Within a TS file I create a very simple `symbol` and an injectable class. When this symbol is used as an injection reference it breaks:

```ts
  const MyGlobalService = Symbol('MyGlobalService');

  // ...
  constructor(
    @Inject(MyGlobalService) private _service: any
  ){}
  //...
```

Error Thrown:
```
Unhandled Promise rejection: Cannot create property '__anonymousType' on symbol 'Symbol(MyGlobalService)' ; Zone: <root> ; Task: Promise.then ; Value: TypeError: Cannot create property '__anonymousType' on symbol 'Symbol(MyGlobalService)'(…) TypeError: Cannot create property '__anonymousType' on symbol 'Symbol(MyGlobalService)'
    at identifierName (https://localhost:8080/lib/@angular/compiler/bundles/compiler.umd.min.js:6:3240)
    at tokenName (https://localhost:8080/lib/@angular/compiler/bundles/compiler.umd.min.js:6:3553)
    at _InjectorBuilder.addProvider (https://localhost:8080/lib/@angular/compiler/bundles/compiler.umd.min.js:34:21529)
    at eval (https://localhost:8080/lib/@angular/compiler/bundles/compiler.umd.min.js:34:19932)
    at Array.forEach (native)
    at NgModuleCompiler.compile (https://localhost:8080/lib/@angular/compiler/bundles/compiler.umd.min.js:34:19890)
    at JitCompiler._compileModule (https://localhost:8080/lib/@angular/compiler/bundles/compiler.umd.min.js:45:25933)
    at createResult (https://localhost:8080/lib/@angular/compiler/bundles/compiler.umd.min.js:45:24480)
    at ZoneDelegate.invoke (https://localhost:8080/lib/zone.js/dist/zone.js:232:26)
    at Zone.run (https://localhost:8080/lib/zone.js/dist/zone.js:114:43)
```

I'm pretty sure the [`identifierName method in compile_metadata.ts`](https://github.com/angular/angular/blob/5614c4ff0fbfc7e856849d1f539934f46b4bca4b/modules/%40angular/compiler/src/compile_metadata.ts#L85-L105) does not handle symbols.

**Expected behavior**
Previously this was working in 2.1. I did try changing the `Symbol` to an `OpaqueToken`, which did fix the issue. Regardless, this is a breaking change that should be noted.

**Minimal reproduction of the problem with instructions**

*MINIMAL DEMO:* http://plnkr.co/edit/6J3klFpWH4Fn1hHna8zo?p=preview

*STEPS TO REPRODUCE:*
 1. Create a `Symbol`.
 1. Create a factory provider using the `Symbol` that returns any object.
 1. Create a component with a constructor that `@Inject`'s using the `Symbol`.
 1. Load the app in Chrome.

**What is the motivation / use case for changing the behavior?**
I am using a `Symbol` to inject a non-angular object. I understand now that an `OpaqueToken` is the recommended approach, but this has broken since an update.

**Please tell us about your environment:**

Windows 8.1, VS Code, NPM, dotnet core (Kestrel)

* **Angular version:** 2.3.0

* **Browser:** [ Chrome 54 ] (works fine in IE 11)

 
* **Language:** [ TypeScript 2.1 ]

* **Node (for AoT issues):** `node --version` =   v6.3.0

### PR DESCRIPTION
**Pull request by DzmitryShylovich**
Closes https://github.com/angular/angular/issues/13314

### CODE REVIEW

File: modules/@angular/compiler/src/compile_metadata.ts [hunk 1]
```diff
@@ -90,6 +90,9 @@ export function identifierName(compileIdentifier: CompileIdentifierMetadata): st
   if (ref instanceof StaticSymbol) {
     return ref.name;
   }
+  if (typeof ref === 'symbol') {
+    return ref.toString().replace(/\(|\)/g, '_');
```
Comment (vicb) [inline modules/@angular/compiler/src/compile_metadata.ts]: But `Symbol("foo") !== Symbol("foo")` - we should probably use a `(Weak)Map` ?

Comment (DzmitryShylovich) [inline modules/@angular/compiler/src/compile_metadata.ts]: > But Symbol("foo") !== Symbol("foo")

we can add a counter to ensure uniqueness `${ref.toString().replace(/\(|\)/g, '_')}${counter++}`

>  use a (Weak)Map ?

what will be the key?

Comment (vicb) [inline modules/@angular/compiler/src/compile_metadata.ts]: > we can add a counter to ensure uniqueness

Then you could not override the token, no ? (same Symbol will generate 2 different IDs)

 > what will be the key?

the Symbol instance ?

File: modules/@angular/compiler/src/compile_metadata.ts [hunk 2]
```diff
@@ -90,6 +92,14 @@ export function identifierName(compileIdentifier: CompileIdentifierMetadata): st
   if (ref instanceof StaticSymbol) {
     return ref.name;
   }
+  if (isSymbol(ref)) {
+    if (symbols.has(ref)) {
+      return symbols.get(ref);
+    }
+    const symbolStr = `${ref.toString().replace(/\(|\)/g, '_')}${symbolIndex++}`;
```
Comment (vicb) [inline modules/@angular/compiler/src/compile_metadata.ts]: could you make it `_symbol_${_sanitizeIdentifier(ref.toString())}_${symbolIndex++}`

File: modules/@angular/compiler/test/integration_spec.ts [hunk 3]
```diff
@@ -38,6 +38,23 @@ export function main() {
...
+         const SOME_SYMBOL = Symbol('SomeSymbol');
+
+         @Component({selector: 'symbol', template: ''})
+         class CmpWithSymbol {
+           constructor(@Inject(SOME_SYMBOL) public symbol: string) {}
+         }
+
+         TestBed.configureTestingModule({
+           declarations: [CmpWithSymbol],
+           providers: [{provide: SOME_SYMBOL, useValue: 'value'}]
```
Comment (vicb) [inline modules/@angular/compiler/test/integration_spec.ts]: and an override ?

**Comment by DzmitryShylovich** [comment]
@vicb done

File: modules/@angular/compiler/test/integration_spec.ts [hunk 4]
```diff
@@ -38,6 +38,30 @@ export function main() {
...
+         class CmpWithSymbol {
+           constructor(
+               @Inject(SOME_SYMBOL) public symbol: string,
+               @Inject(ANOTHER_SYMBOL) public anotherSymbol: string) {}
+         }
+
+         TestBed.configureTestingModule({
+           declarations: [CmpWithSymbol],
+           providers: [
+             {provide: SOME_SYMBOL, useValue: 'value'},
```
Comment (vicb) [inline modules/@angular/compiler/test/integration_spec.ts]:
```
{provide: SOME_SYMBOL, useValue: 'value'},
{provide: SOME_SYMBOL, useValue: 'override'},
```

**Comment by DzmitryShylovich** [comment]
@vicb comment is addressed

File: modules/@angular/compiler/src/compile_metadata.ts [hunk 5]
```diff
@@ -81,6 +81,8 @@ function _sanitizeIdentifier(name: string): string {
 }
 
 let _anonymousTypeIndex = 0;
+let symbolIndex = 0;
+const symbols = new Map<Symbol, string>();
```
Comment (vicb) [inline modules/@angular/compiler/src/compile_metadata.ts]: please rename to `symboIds`

File: modules/@angular/compiler/src/compile_metadata.ts [hunk 6]
```diff
@@ -90,6 +92,14 @@ export function identifierName(compileIdentifier: CompileIdentifierMetadata): st
   if (ref instanceof StaticSymbol) {
     return ref.name;
   }
+  if (isSymbol(ref)) {
+    if (symbols.has(ref)) {
+      return symbols.get(ref);
+    }
+    const symbolStr = `_symbol_${_sanitizeIdentifier(ref.toString())}_${symbolIndex++}`;
```
Comment (vicb) [inline modules/@angular/compiler/src/compile_metadata.ts]: and `symbolId`

**Comment by DzmitryShylovich** [comment]
@vicb done

**Review by mhevery — APPROVED** [review]

**Comment by maximusk** [comment]
Is Symbol essentially a replacement OpaqueToken ?

**Comment by DzmitryShylovich** [comment]
nope

**Comment by maximusk** [comment]
It would be interesting to see what OpaqueToken can do that Symbol can't

**Pull request merged**